freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingService.java

```diff
@@ -0,0 +1,122 @@
+import com.sequenceiq.freeipa.flow.freeipa.upscale.UpscaleFlowEvent;
+import com.sequenceiq.freeipa.flow.freeipa.upscale.event.UpscaleEvent;
+import com.sequenceiq.freeipa.service.freeipa.flow.FreeIpaFlowManager;
+import com.sequenceiq.freeipa.service.operation.OperationService;
+
+@Service
+public class FreeIpaScalingService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaScalingService.class);
+
+    @Inject
+    private OperationService operationService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private FreeIpaFlowManager flowManager;
+
+    @Inject
+    private FreeIpaScalingValidationService validationService;
+
+    public UpscaleResponse upscale(String accountId, UpscaleRequest request) {
+        Stack stack = getStack(accountId, request.getEnvironmentCrn());
+        Set<InstanceMetaData> allInstances = stack.getNotDeletedInstanceMetaDataSet();
+        FormFactor originalFormFactor = FormFactor.getByInstanceCount(allInstances.size());
+        validationService.validateStackForUpscale(allInstances, stack, new ScalingPath(originalFormFactor, request.getTargetFormFactor()));
+        return triggerUpscale(request, stack, originalFormFactor);
+    }
+
+    public DownscaleResponse downscale(String accountId, DownscaleRequest request) {
+        Stack stack = getStack(accountId, request.getEnvironmentCrn());
+        Set<InstanceMetaData> allInstances = stack.getNotDeletedInstanceMetaDataSet();
+        FormFactor originalFormFactor = FormFactor.getByInstanceCount(allInstances.size());
+        validationService.validateStackForDownscale(allInstances, stack, new ScalingPath(originalFormFactor, request.getTargetFormFactor()));
+        return triggerDownscale(request, stack, originalFormFactor);
+    }
+
+    private Stack getStack(String accountId, String envCrn) {
+        Stack stack = stackService.getByEnvironmentCrnAndAccountIdWithLists(envCrn, accountId);
+        MDCBuilder.buildMdcContext(stack);
+        return stack;
+    }
+
+    private UpscaleResponse triggerUpscale(UpscaleRequest request, Stack stack, FormFactor originalFormFactor) {
+        Operation operation = startScalingOperation(stack.getAccountId(), request.getEnvironmentCrn(), OperationType.UPSCALE);
+        UpscaleEvent upscaleEvent = new UpscaleEvent(UpscaleFlowEvent.UPSCALE_EVENT.event(),
+                stack.getId(), request.getTargetFormFactor().getInstanceCount(), false, false, false, operation.getOperationId());
+        LOGGER.info("Trigger upscale flow with event: {}", upscaleEvent);
+        FlowIdentifier flowIdentifier = flowManager.notify(UpscaleFlowEvent.UPSCALE_EVENT.event(), upscaleEvent);
+        UpscaleResponse response = new UpscaleResponse();
+        response.setOperationId(operation.getOperationId());
+        response.setOriginalFormFactor(originalFormFactor);
+        response.setTargetFormFactor(request.getTargetFormFactor());
+        response.setFlowIdentifier(flowIdentifier);
```

Comment:
scale response creation could be extracted into a common method since both response classes are childs of ScaleResponseBase

Reply:
I wouldn't over-generalize this logic if response objects would diverge in the future, we already have a follow-up to accept instanceIds for downscale which does not make sense for upscale

Reply:
ok, it makes sense